### ISSUE TITLE
Performance problem with product collection (media/gallery data)

### ISSUE THREAD

**Issue by redboxmarcins** [magento/magento2#6850]
### Preconditions

1. Present in all versions (2.0.x and 2.1.x)
2. Easy to replicate, but problem appeared as we have over 100k products
### Steps to reproduce

1. Enable mysql slow query log or use profiler
2. Visit any product listing page
3. Observe one query appearing over and over again
### Expected result

1. Page loads quickly and there are no large repeated queries
### Actual result

1. There is one query executed over and over again

 Collection is adding media data here - https://github.com/magento/magento2/blob/2.1/app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php#L2207
 which is executed in that class https://github.com/magento/magento2/blob/308da7d18a46d2a3e4dc8a553e38f5ec58b476f4/app/code/Magento/Catalog/Model/ResourceModel/Product/Gallery.php#L150

 and this method will load ALL gallery entries per store 

My proposal is to change that method so it can accept additional parameter where you can limit that to certain product ids

f.e.

```
$select = $this->getMediaGalleryResource()->createBatchBaseSelect(
            $this->getStoreId(),
            $attribute->getAttributeId(),
            $this->getAllIds()
        );
```

and then

```
public function createBatchBaseSelect($storeId, $attributeId, array $ids = array())
{
...
        if (!empty($ids)) {
            $select->where(
                '`entity`.`entity_id` IN (?)',
                $ids
            );
        }
}
```

Same thing applies to configurable products collection

**Comment by kandy** [comment]
Internal tocket https://jira.corp.magento.com/browse/MAGETWO-58320

**Comment by thlassche** [comment]
@kandy @redboxmarcins This is already fixed in: 7368be9586caa11b23000e26701b65a9df3e2e38

Jira ticket: MAGETWO-59240

**Comment by pmathbliss** [comment]
I found similar problems with this query. I commented out vendor/magento/module-configurable-product/Model/Product/Type/Configurable.php line 589
`$collection->addMediaGalleryData();`
I was then able to open up the admin and product page loads returned a to reasonable speed.

**Comment by magento-engcom-team** [comment]
@redboxmarcins, thank you for your report.
The issue is already fixed in 2.1.9
